components/Input.tsx

```diff
@@ -1,41 +1,126 @@
 import Image from "next/image";
 import { useState } from "react";
+import {
+  FieldErrors,
+  FieldValues,
+  UseFormGetValues,
+  UseFormRegister,
+  UseFormSetError,
+} from "react-hook-form";
+import { postEmailIsValid } from "@/lib/api";
 import eyeOff from "@/public/eye-off.png";
 import eyeOn from "@/public/eye-on.png";
+import { InputType } from "@/types/commonTypes";
 import styles from "./Input.module.css";
 
-function Input() {
-  const [isError, setIsError] = useState(false);
-  const [isEyeOn, setIsEyeOn] = useState(false);
+interface Props {
+  placeholder: string;
+  type: InputType;
+  register: UseFormRegister<FieldValues>;
+  errors: FieldErrors<FieldValues>;
+  setError: UseFormSetError<FieldValues>;
+  getValues: UseFormGetValues<FieldValues>;
+}
+
+const options = {
+  signInEmail: {
+    required: "이메일을 입력해주세요",
+  },
+  signInPassword: {
+    required: "비밀번호를 입력해주세요",
+  },
+  signUpEmail: {
+    required: "이메일을 입력해주세요",
+    validate: {
+      isValidEmail: async (email: FieldValues) => {
+        try {
+          const response = await postEmailIsValid({ email });
+          return response && true;
+        } catch (errorMessage) {
+          return errorMessage;
+        }
+      },
+    },
+  },
+  signUpPassword: {
+    required: "비밀번호를 입력해주세요",
+    minLength: {
+      value: 8,
+      message: "비밀번호는 영문, 숫자 조합 8자 이상 입력해 주세요",
+    },
+  },
+};
+
+const Type = {
+  email: ["signInEmail", "signUpEmail"],
+  password: ["signInPassword", "signUpPassword", "signUpPasswordCheck"],
+};
+
+function Input({ placeholder, type, register, errors, getValues }: Props) {
+  const myOptions = {
+    ...options,
+    signUpPasswordCheck: {
+      validate: {
+        isSamePassword: (passwordCheck: string) => {
+          const values = getValues();
+          const password = values["signUpPassword"];
+          return passwordCheck === password || "비밀번호가 일치하지 않아요";
+        },
+      },
+    },
+  };
+  const { ...rest } = register(type, myOptions[type]);
+  const [isPasswordVisible, SetIsPasswordVisible] = useState(false);
 
-  const inputClassName = isError
+  const inputClassName = errors[type]?.message
     ? `${styles.input} ${styles.error}`
     : styles.input;
```

Comment:
조건부로 변경되는 부분에만 조건을 따로 적용하시는 게 유지보수에 좋습니다.
```suggestion
  const inputClassName = `${styles.input}${errors[type]?.message ? ` ${styles.error}` : ''}`;
```